### PR TITLE
delete in stream telemetry - in demo + test

### PR DESCRIPTION
**Pull request by micnovak**
minor refactoring
stream telemetry delete in demo adapter
tests for delete in stream telemetry

### CODE REVIEW

File: tests/client_server_test_base.py [hunk 1]
```diff
@@ -530,15 +605,38 @@ def test_subscribe_stream(self, request, data_type):
         self._test_subscribe(prefix_str, self.NS_IETF_INTERFACES,
                              paths, changes_list)
 
+    @pytest.mark.usefixtures("reset_cfg")
+    def test_subscribe_stream_delete(self, request):
+        log.info("testing subscribe_stream_delete")
+
+        changes_list = [
+            ("top-d", ChangeDel(deleted_paths=[
```
Comment (martin-volf) [inline tests/client_server_test_base.py:613]: You are introducing more and more dependencies on demo adapter classes.

File: tests/client_server_test_base.py [hunk 2]
```diff
@@ -454,34 +508,50 @@ def confd_cmd_subprocess(confd_cmd):
...
+                    pv_res.append(pv_candidates)
+                    del_res.append(del_candidates)
+                    pv_candidates, del_candidates = [], []
             else:
-                if len(path_value) < pv_idx + 1:
-                    path_value.append([])
-                path_value[pv_idx].append((make_gnmi_path(c[0]), c[1]))
-        log.debug("path_value=%s", path_value)
-        return path_value
+                if isinstance(c[1], (str, int)) or isinstance(c[1], ChangeVal):
```
Comment (martin-volf) [inline tests/client_server_test_base.py:543]: I don't see where `ChangeVal` would be created, is it really needed?

File: tests/client_server_test_base.py [hunk 3]
```diff
@@ -548,13 +646,16 @@ def _test_subscribe(self, prefix_str, ns_prefix, paths, changes_list):
             GnmiDemoServerAdapter.load_config_string(
                 self._changes_list_to_xml(changes_list, prefix_pfx))
         if self.adapter_type == AdapterType.API:
-            prefix_pfx = prefix_str.format(prefix=self.PREFIX_MAP[ns_prefix])
+            if ns_prefix in self.PREFIX_MAP:
+                prefix_pfx = prefix_str.format(prefix=self.PREFIX_MAP[ns_prefix])
+            else:
+                prefix_pfx = prefix_str.format(prefix='')
```
Comment (martin-volf) [inline tests/client_server_test_base.py:652]: or:
```python
prefix_pfx = prefix_str.format(prefix=self.PREFIX_MAP.get(ns_prefix, ''))
```

**Review by martin-volf — COMMENTED** [review]
The base test classes contain *a lot* of dependencies on the demo adapter (there were some already, but this change introduces a lot more). It is "just" a test, so I guess it is acceptable, but still.

File: src/confd_gnmi_demo_adapter.py [hunk 4]
```diff
@@ -297,12 +336,66 @@ def process_changes(self):
                     if add_count == 0:
                         add_count += 1
                         # sometimes skip add if is first from previous
-                        if randint(0, 1) == 0:
-                            continue
+                        if randint(0, 1) == 0: continue
```
[the remark this answers is not in the document]
Comment (jomiklos) [reply src/confd_gnmi_demo_adapter.py:344]: imho pep 8 is the king!

File: tests/client_server_test_base.py [hunk 5]
```diff
@@ -37,6 +45,7 @@ def fix_method(self, request):
         self.set_adapter_type()
         self.server = ConfDgNMIServicer.serve(adapter_type=self.adapter_type, insecure=True)
         self.client = ConfDgNMIClient(insecure=True)
+        GnmiDemoServerAdapter.fill_demo_db()
```
[the remark this answers is not in the document]
Comment (micnovak) [reply tests/client_server_test_base.py:48]: demo db is currently also used for ConfD tests (contains paths to test)

File: tests/client_server_test_base.py (hunk 1 above)
Comment (micnovak) [reply tests/client_server_test_base.py:613]: It has inner method :-)

File: src/confd_gnmi_demo_adapter.py [hunk 6]
```diff
@@ -86,29 +124,49 @@ def get_adapter(cls):
...
+                state_if_name = f"state_if_{if_id}"
+                path = f"/interfaces/interface[name={if_name}]"
+                state_path = f"/interfaces-state/interface[name={state_if_name}]"
+                cls.demo_db[f"{path}/name"] = if_name
+                cls.demo_state_db[f"{state_path}/name"] = state_if_name
+                cls.demo_db[f"{path}/type"] = "gigabitEthernet"
+                cls.demo_state_db[f"{state_path}/type"] = "gigabitEthernet"
+                cls.demo_db[f"{path}/enabled"] = True
+
+            for i in range(1, 5):
```
Comment (jomiklos) [inline src/confd_gnmi_demo_adapter.py:148]: does loop make sense at all here if every instance is different?
maybe just do simple loop unwrap...?

File: tests/client_server_test_base.py (hunk 2 above)
Comment (micnovak) [reply tests/client_server_test_base.py:543]: Not now, it was added with ChangeDel so we have type for each operation. More correctly `str` should not be used but I kept it

File: gnmi-tools.xml [hunk 7]
```diff
@@ -54,5 +54,25 @@
             <type>opticalTransport</type>
             <admin-state>In-Service</admin-state>
         </double-key-list>
+        <top-d>
```
[the remark this answers is not in the document]
Comment (micnovak) [reply gnmi-tools.xml:57]: top for `delete` functionality testing

File: src/confd_gnmi_adapter.py [hunk 8]
```diff
@@ -125,11 +125,10 @@ def add_path_for_monitoring(self, path, prefix):
             pass
 
         @abstractmethod
-        def get_monitored_changes(self) -> List:
+        def get_subscription_notifications(self):
```
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_adapter.py:128]: I have removed full function and took header from API adapter, which did not contain it. The returned value is now different, even though it is still `List`.

File: tests/client_server_test_base.py [hunk 9]
```diff
@@ -258,25 +305,30 @@ def read_subscribe_responses(responses, read_count=-1):
...
+            read_fun=read_fun,
+            poll_interval=poll_interval,
+            poll_count=poll_count,
+            read_count=read_count
+        )
 
         log.debug("responses=%s", responses)
         if poll_count:
-            assert (poll_count + 1 == response_count)
+            assert poll_count + 1 == response_count
```
[the remark this answers is not in the document]
Comment (micnovak) [reply tests/client_server_test_base.py:329]: It has inner method :-)

File: tests/client_server_test_base.py (hunk 1 above)
Comment (micnovak) [reply tests/client_server_test_base.py:613]: This class can be used here in tests as well.

File: tests/client_server_test_base.py (hunk 3 above)
Comment (micnovak) [reply tests/client_server_test_base.py:652]: ok

File: src/confd_gnmi_adapter.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (micnovak) [reply src/confd_gnmi_adapter.py:128]: ok

File: src/confd_gnmi_demo_adapter.py (hunk 6 above)
Comment (micnovak) [reply src/confd_gnmi_demo_adapter.py:148]: ok

**Review by jomiklos — APPROVED** [review]

**Pull request merged**